sklearn/ensemble/_forest.py

```diff
@@ -597,16 +616,19 @@ def _compute_oob_predictions(self, X, y):
             n_samples,
             self.max_samples,
         )
-        for estimator in self.estimators_:
-            unsampled_indices = _generate_unsampled_indices(
-                estimator.random_state,
-                n_samples,
-                n_samples_bootstrap,
-            )
 
-            y_pred = self._get_oob_predictions(estimator, X[unsampled_indices, :])
-            oob_pred[unsampled_indices, ...] += y_pred
-            n_oob_pred[unsampled_indices, :] += 1
+        lock = threading.Lock()
+        sub_ests = []
+        chunk_size = 40
```

Comment:
@glemaitre @thomasjpfan I've refactored the parallel strategy to reduce the amount of temporary data that gets stored while avoiding severe lock contention by using chunking on top of a lock (locking alone abrogates all performance gains--it seems that giving each thread a "chunk" of work reduces contention vs. whatever `joblib` does by default). I'm not particularly fond of this solution (the magic number here is pretty atrocious), but it does provide clear wins in the `asv` benchmarks that were previously showing tradeoffs on memory consumption.

Repeating the benchmarks from https://github.com/scikit-learn/scikit-learn/pull/28442#issuecomment-2334910902 (on the same machine):

```
export SKLBENCH_NJOBS=[1,8]
time asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059
```

```
| Change   | Before [429d67aa] <main>   | After [2325f154] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                |
|----------|----------------------------|-----------------------------------------|---------|----------------------------------------------------------------------|
| -        | 1.10±0.04s                 | 910±6ms                                 |    0.83 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 8, True) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

```
export SKLBENCH_PROFILE=large_scale
export SKLBENCH_NJOBS=[16]
time asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059
```

```
| Change   | Before [429d67aa] <main>   | After [2325f154] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                 |
|----------|----------------------------|-----------------------------------------|---------|-----------------------------------------------------------------------|
| -        | 3.56±0.01s                 | 2.50±0.06s                              |     0.7 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 16, True) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

I'm not convinced my time was well spent here though--I tried a lot of different things, and at the end of the day it seems like what we really need is a highly-performant "atomic accumulator" data structure where we can quickly accumulate/reduce the predictions without unreasonable lock contention nor usage of a "magic number" for chunking.

Also, the `asv` benchmarks for the project may not necessarily provide a solid proxy for our in-house use case (ratio of features:records around 900_000:900).  If I repeat the informal scaling script from the original ticket on the latest version of this branch, while more impressive from the `asv` standpoint, it looks much less impressive for the original real-world use case proxy:

<details>

![image](https://github.com/user-attachments/assets/cfc21ff2-fe5d-4839-9eb2-c0d188b6c770)


</details>

That might be a little better than before, but we'd likely still have to vendor our own solution for the foreseeable future.

It seems like there are at least two routes forward:

1) Accept that the `asv` benchmarks aren't particularly great proxies for what we're trying to accomplish, and accept the memory costs for the infrastructure changes that perform well in the "real world." You do have a separate non-asv `benchmarks` folder, so this may be a situation you are not entirely unfamiliar with. That said, I would agree that it would be really nice if faster `asv` proxy benchmarks could do the trick. Perhaps with a much larger feature:record ratio they might.
2) Continue to use `asv` as the "gatekeeper"/standard, and accept that for now this is a step toward using parallel OOB, but that what we really need is a solution with both reasonable memory footprint and excellent real-world scaling behavior.

So, it would probably be good to decide what path you'd like to take here.